### ISSUE TITLE
"vscode.TreeItemCollapsible" Never Changes

### ISSUE THREAD

**Issue by rroessler** [eclipse-theia/theia#13946]
### Bug Description:

In VSC extensions, whenever a refresh is requested by a `vscode.TreeDataProvider`, trees are refreshed to an "initialized" collapsible state instead of maintaining their current collapsible state.

This is an issue for tree-views that show real-time data (eg: custom debugger panels) as refreshes will force all nodes to snap back to their original collapsible state.

### Steps to Reproduce:

This issue can be reproduced by implementing a simple tree-view provider that refreshes every 5-seconds. If the root-node is clicked, anytime the interval fires the root-node will reset to the original collapsed state as opposed to adhering to the current expanded state.

```typescript
import vscode from 'vscode';

class Provider implements vscode.TreeDataProvider {
    private m_root = new vscode.TreeItem('Root', vscode.TreeItemCollapsibleState.Collapsed);
    private m_children = Array.from(Array(5), (_, ii) => new vscode.TreeItem(`Node: ${ii}`));

    readonly emitter = new vscode.EventEmitter<vscode.TreeItem | vscode.TreeItem[] | undefined | null | void>();
    readonly onDidChangeTreeData = this.emitter.event.bind(this.emitter);

    getTreeItem(element: vscode.TreeItem): vscode.TreeItem {
        return element;
    }

    getChildren(element?: vscode.TreeItem | undefined): vscode.ProviderResult<vscode.TreeItem[]> {
        if (typeof element === 'undefined') return [this.m_root];
        return element === this.m_root ? this.m_children : [];
    }
}

export function activate(context: vscode.ExtensionContext) {
    const treeDataProvider = new Provider();
    context.subscriptions.push(vscode.window.createTreeView('exampleTreeProvider', { treeDataProvider }));

    // set an interval to trigger refreshes every 5 seconds
    const interval = setInterval(() => treeDataProvider.emitter.fire(), 5000);
    context.subscriptions.push(() => clearInterval(interval));
}
```

**Comment by rroessler** [comment]
Alongside this, triggering refreshes via "onDidChangeTreeData" causes the current scroll-state to reset to an initial position.

**Comment by tsmaeder** [comment]
@rroessler do you know what the behavior in VS Code is?

**Comment by rroessler** [comment]
@tsmaeder The original behaviour in VS Code is that when `onDidChangeTreeData` is called with specific node as arguments, _only_ those nodes are refreshed. This means that parent nodes would not be refreshed, maintaining their current expansion state.

Currently the Theia extension wrapper forces a refresh on the entire `TreeView` widget. As such, nodes that may have been created with a collapsed state (eg: at `getChildren` or `getTreeItem`) but where then expanded before a refresh occurs, are all refreshed causing flickering/odd effects if refreshing occurs frequently.

Alongside this, the VS Code behaviour has the added bonus of being more efficient as only the nodes requested are refreshed, as opposed to the Theia implementation of refresh all from the root.

**Comment by tsmaeder** [comment]
@rroessler just so we're talking about the same thing: the example extension you give refreshes the root of the tree, though, right?

**Comment by rroessler** [comment]
@tsmaeder you are correct as the issue also arises from here too. The difference between the two behaviours stems from:

VS Code:
- When a tree-item "change" event occurs, only that node and its children are refreshed.
- Additionally, tree-items are have an associated state that is maintained between refreshes (eg: expansion, scroll-top).

Theia:
- When a tree-item "change" event occurs, the entire tree is refreshed/rebuilt.
- This means that every node is recreated from their source and the associated state is not maintained between refreshes.

**Comment by tsmaeder** [comment]
@rroessler since you seem to know how what to do, would you be ready to contribute the fix?

**Comment by rroessler** [comment]
@tsmaeder I will try my best to get a fix soon.

**Comment by abdelhadyahmed** [comment]
Hello @rroessler ,
Is this issue related to treeview getting rebuilt and collapsed if any change happens to the treeview ?
I'm just asking because I'm facing the same issue.
I have a normal behavior on vscode extension but in theia it collapse all nodes in rebuild.

**Comment by tsmaeder** [comment]
@rroessler I opened https://github.com/eclipse-theia/theia/pull/14697
